modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java

```diff
@@ -2163,68 +2159,44 @@ public void repopulate(final ElasticsearchIndex index) throws IndexRebuildExcept
       logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
       throw new IndexRebuildException(index.getIndexName(), getService(), e);
     }
-    final int limit = 1000;
 
     if (total > 0) {
       logIndexRebuildBegin(logger.getSlf4jLogger(), index.getIndexName(), total, "workflows");
+      List<WorkflowIndexData> workflowIndexData;
+
+      int limit = 1000;
+      int offset = 0;
       int current = 0;
-      long token = 0;
-      List<WorkflowInstance> workflows;
+      String currentMediapackageId;
+      String lastMediapackageId = "";
+
       do {
         try {
-          workflows = persistence.getLatestWorkflowInstances(limit, token);
-        } catch (Exception e) {
-          logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), total, current, e);
+          workflowIndexData = persistence.getWorkflowIndexData(limit, offset);
+        } catch (WorkflowDatabaseException e) {
+          logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
           throw new IndexRebuildException(index.getIndexName(), getService(), e);
         }
-        logger.debug("Got {} workflows for re-indexing", workflows.size());
-        int indexLast = workflows.size() - 1;
-        token = workflows.get(indexLast).getId();
+        offset += limit;
+        logger.debug("Got {} workflows for re-indexing", workflowIndexData.size());
 
-        for (WorkflowInstance instance : workflows) {
-          current += 1;
-          Organization organization = null;
-          try {
-            organization = organizationDirectoryService.getOrganization(instance.getOrganizationId());
-          } catch (NotFoundException e) {
-            logger.error("Found workflow with non-existing organization {}", instance.getOrganizationId());
+        for (WorkflowIndexData data : workflowIndexData) {
+          currentMediapackageId = data.getMediaPackageId();
+          if (currentMediapackageId.equals(lastMediapackageId)) {
             continue;
           }
+          else {
```

Comment:
Else is unnecessary since you already terminate the loop with continue. Else invert the confition and do not use continue.
```suggestion
```